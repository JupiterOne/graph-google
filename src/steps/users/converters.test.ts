import { getMockUser } from '../../../test/mocks';
import {
  createUserEntity,
  getCollectionAsFlattendFields,
  convertCustomSchemas,
} from './converters';

describe('#createUserEntity', () => {
  test('should convert to entity', () => {
    expect(createUserEntity(getMockUser())).toMatchSnapshot();
  });
});

describe('#getCollectionAsFlattenedFields', () => {
  getCollectionAsFlattendFields({
    collection: [
      {
        type: 'work',
      },
    ],
    suffix: 'Address',
    valueMethod: 'formatted',
  });
});

describe('#convertCustomSchemas', () => {
  const original: any = {
    string: 'a',
    array: ['a', 'b', 'c'],
    number: 123,
    aGoodTime: '2019-04-23T18:06:05Z',
    another_time: '2019-04-23T18:06:05Z',
    someDate: '2019-04-23T18:06:05Z',
    occurredOn: '2019-04-23T18:06:05Z',
    updatedAt: '2019-04-23T18:06:05Z',
    aBadTime: 'do I look like time to you?',
    anUndefinedTime: undefined,
    aNullTime: null,
    name: 'you',
    object: {
      name: 'me',
      firstLevel: {
        name: 'me again',
        secondLevel: {
          thirdLevel: 'hey',
        },
      },
      array: [
        {
          k: 'v',
        },
        {
          k: 'v',
        },
      ],
    },
    objectArray: [
      {
        wut: 'no matter',
      },
    ],
    arrayOfNull: [null],
    arrayOfUndefined: [undefined],
  };

  const converted: any = {
    string: 'a',
    array: ['a', 'b', 'c'],
    number: 123,
    aGoodTime: '2019-04-23T18:06:05Z',
    another_time: '2019-04-23T18:06:05Z',
    someDate: '2019-04-23T18:06:05Z',
    occurredOn: '2019-04-23T18:06:05Z',
    updatedAt: '2019-04-23T18:06:05Z',
    aBadTime: 'do I look like time to you?',
    name: 'you',
    'object.name': 'me',
    'object.firstLevel.name': 'me again',
    'object.firstLevel.secondLevel.thirdLevel': 'hey',
    'object.array': original.object.array.map(
      (a) => typeof a === 'object' && JSON.stringify(a),
    ),
    objectArray: original.objectArray.map(
      (a) => typeof a === 'object' && JSON.stringify(a),
    ),
    arrayOfNull: [null],
    arrayOfUndefined: [undefined],
    anUndefinedTime: undefined,
    aNullTime: null,
  };

  const originalWithoutPrefix: any = {
    one: 'i',
    two: 'am',
    three: {
      four: 'test',
    },
  };

  const convertedWithPrefix: any = {
    'test.one': 'i',
    'test.two': 'am',
    'test.three.four': 'test',
  };

  test('default options', () => {
    expect(convertCustomSchemas(original)).toEqual(converted);
  });

  test('expected prefix', () => {
    expect(convertCustomSchemas(originalWithoutPrefix, 'test')).toEqual(
      convertedWithPrefix,
    );
  });
});
