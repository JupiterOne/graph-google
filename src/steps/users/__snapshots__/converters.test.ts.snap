// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createUserEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "google_user_123456789",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "agreedToTerms": true,
        "archived": false,
        "changePasswordAtNextLogin": false,
        "creationTime": "2020-08-13T19:03:11.000Z",
        "customSchemas": Object {
          "SSO": Object {
            "role": Array [
              Object {
                "type": "work",
                "value": "arn:aws:iam::123456789:role/Developer,arn:aws:iam::123456789:saml-provider/gsuite",
              },
            ],
          },
        },
        "customerId": "abc123",
        "emails": Array [
          Object {
            "address": "john.doe@jupiterone.io",
            "primary": true,
            "type": "work",
          },
        ],
        "etag": "abcdef",
        "id": "123456789",
        "includeInGlobalAddressList": true,
        "ipWhitelisted": false,
        "isAdmin": false,
        "isDelegatedAdmin": false,
        "isEnforcedIn2Sv": true,
        "isEnrolledIn2Sv": true,
        "isMailboxSetup": true,
        "kind": "admin#directory#user",
        "lastLoginTime": "2020-08-27T00:57:59.000Z",
        "locations": Array [
          Object {
            "area": "desk",
            "buildingId": "RTP",
            "type": "desk",
          },
        ],
        "name": Object {
          "familyName": "Doe",
          "fullName": "John Doe",
          "givenName": "John",
        },
        "orgUnitPath": "/Marketing",
        "organizations": Array [
          Object {
            "costCenter": "01A",
            "customType": "",
            "department": "Marketing",
            "description": "Full-time",
            "primary": true,
            "title": "CMO",
          },
        ],
        "primaryEmail": "john.doe@jupiterone.io",
        "recoveryEmail": "john.doe.recovery@jupiterone.io",
        "recoveryPhone": "+19999999999",
        "relations": Array [
          Object {
            "type": "manager",
            "value": "john.doe.manager@jupiterone.io",
          },
        ],
        "suspended": false,
      },
    },
  ],
  "_type": "google_user",
  "active": true,
  "admin": false,
  "agreedToTerms": true,
  "aliases": undefined,
  "archived": false,
  "changePasswordAtNextLogin": false,
  "costCenter": "01A",
  "createdOn": 1597345391000,
  "creationTime": 1597345391000,
  "customSchemas": "{\\"SSO\\":{\\"role\\":[{\\"type\\":\\"work\\",\\"value\\":\\"arn:aws:iam::123456789:role/Developer,arn:aws:iam::123456789:saml-provider/gsuite\\"}]}}",
  "customType": "",
  "customerId": "abc123",
  "deletedOn": undefined,
  "deletionTime": undefined,
  "department": "Marketing",
  "description": "Full-time",
  "displayName": "John Doe",
  "email": "john.doe@jupiterone.io",
  "emailDomain": Array [
    "jupiterone.io",
  ],
  "employeeType": "Full-time",
  "firstName": "John",
  "hashFunction": undefined,
  "id": "123456789",
  "includeInGlobalAddressList": true,
  "ipWhitelisted": false,
  "isAdmin": false,
  "isDelegatedAdmin": false,
  "isEnforcedIn2Sv": true,
  "isEnrolledIn2Sv": true,
  "isMailboxSetup": true,
  "kind": "admin#directory#user",
  "lastLoginOn": 1598489879000,
  "lastLoginTime": 1598489879000,
  "lastName": "Doe",
  "managerEmail": "john.doe.manager@jupiterone.io",
  "managerRelation": "john.doe.manager@jupiterone.io",
  "mfaEnabled": true,
  "name": "John Doe",
  "orgUnitPath": "/Marketing",
  "primaryEmail": "john.doe@jupiterone.io",
  "recoveryEmail": "john.doe.recovery@jupiterone.io",
  "recoveryPhone": "+19999999999",
  "suspended": false,
  "suspensionReason": undefined,
  "thumbnailPhotoUrl": undefined,
  "title": "CMO",
  "username": "john.doe@jupiterone.io",
  "workEmail": "john.doe@jupiterone.io",
}
`;
